plugins/goodix-tp/fu-goodixtp-brlb.c

```diff
@@ -0,0 +1,519 @@
+		ret = read_pkg(device, tmp_addr + offset, &buf[offset], pkg_size, error);
+		if (!ret)
+			return ret;
+		offset += pkg_size;
+	}
+
+	if (remain_size > 0) {
+		ret = read_pkg(device, tmp_addr + offset, &buf[offset], remain_size, error);
+		if (!ret)
+			return ret;
+	}
+	return TRUE;
+}
+
+static gboolean
+hid_write(FuDevice *device, guint32 addr, guint8 *buf, guint32 len, GError **error)
+{
+	guint8 HidBuf[PACKAGE_LEN] = {0};
+	guint32 current_addr = addr;
+	guint32 transfer_length = 0;
+	guint32 pos = 0;
+	guint8 pkg_num = 0;
+	gboolean ret;
+
+	while (pos != len) {
+		HidBuf[0] = REPORT_ID;
+		HidBuf[1] = I2C_DIRECT_RW;
+		if (len - pos > PACKAGE_LEN - 12) {
+			transfer_length = PACKAGE_LEN - 12;
+			HidBuf[2] = 0x01;
+		} else {
+			transfer_length = len - pos;
+			HidBuf[2] = 0x00;
+		}
+		HidBuf[3] = pkg_num++;
+		HidBuf[4] = transfer_length + 7;
+		HidBuf[5] = I2C_WRITE_FLAG;
+		HidBuf[6] = (current_addr >> 24) & 0xFF;
+		HidBuf[7] = (current_addr >> 16) & 0xFF;
+		HidBuf[8] = (current_addr >> 8) & 0xFF;
+		HidBuf[9] = current_addr & 0xFF;
+		HidBuf[10] = (transfer_length >> 8) & 0xFF;
+		HidBuf[11] = transfer_length & 0xFF;
+		memcpy(&HidBuf[12], &buf[pos], transfer_length);
+		ret = set_report(device, HidBuf, transfer_length + 12, error);
+		if (!ret) {
+			g_debug("Failed write data to addr=0x%x, len=%d",
+				current_addr,
+				(gint)transfer_length);
+			return ret;
+		}
+		pos += transfer_length;
+		current_addr += transfer_length;
+	}
+	return TRUE;
+}
+
+static gboolean
+send_cmd(FuDevice *device, guint8 cmd, guint8 *data, guint32 dataLen, GError **error)
+{
+	guint8 HidBuf[PACKAGE_LEN] = {0};
+	gboolean ret;
+
+	HidBuf[0] = REPORT_ID;
+	HidBuf[1] = cmd;
+	HidBuf[2] = 0x00;
+	HidBuf[3] = 0x00;
+	HidBuf[4] = (guint8)dataLen;
+	memcpy(&HidBuf[5], data, dataLen);
+	ret = set_report(device, HidBuf, dataLen + 5, error);
+	if (!ret) {
+		g_debug("Failed send cmd 0x%02x", cmd);
+		return ret;
+	}
+	return TRUE;
+}
+
+static gint
+brlb_get_version(FuDevice *device, GError **error)
+{
+	guint8 tempBuf[14] = {0};
+	gint version;
+	guint8 pid[9] = {0};
+	guint8 vid[8] = {0};
+	guint8 vice_ver;
+	guint8 inter_ver;
+	guint8 cfg_ver;
+	guint32 cfg_id;
+	guint8 sensor_id;
+	gboolean ret;
+
+	ret = hid_read(device, 0x1001E, tempBuf, 14, error);
+	if (!ret) {
+		g_debug("Failed read PID/VID");
+		return -1;
+	}
+	memcpy(pid, tempBuf, 8);
+	memcpy(vid, &tempBuf[8], 4);
+	sensor_id = tempBuf[13];
+	vice_ver = tempBuf[10];
+	inter_ver = tempBuf[11];
+
+	ret = hid_read(device, 0x10076, tempBuf, 5, error);
+	if (!ret) {
+		g_debug("Failed read config id/version");
+		return -1;
+	}
+
+	cfg_id = *((guint32 *)tempBuf);
+	cfg_ver = tempBuf[4];
+	version = (vice_ver << 16) | (inter_ver << 8) | cfg_ver;
+
+	g_debug("PID:%s", pid);
+	g_debug("VID:%02x %02x %02x %02x", vid[0], vid[1], vid[2], vid[3]);
+	g_debug("sensorID:%d", sensor_id);
+	g_debug("configID:%08x", cfg_id);
+	g_debug("configVer:%02x", cfg_ver);
+	g_debug("version:%d", version);
```

Comment:
these all need to be saved to GObject private state and then printed in the FuDevice->to_string() helper.